### ISSUE TITLE
.bas obfuscation in APK mode

### ISSUE THREAD

**Issue by mougino** [RFO-BASIC/Basic#181]
As discussed in the forum, here is the evolution I've been working on.

I first added a new boolean value `is_encrypted` in "values/setup.xml" to let the user (or a builder) decide if obfuscation of the .bas file(s) is wanted or not.

Then I modified `Basic.LoadTheProgram()` which is an APK-only method (reads the program from res/raw or assets/<AppPath>/source and puts it into memory). I basically load the whole InputStream into a String, decode it, then re-build a new decoded InputStream from the String.
I made a PC-world equivalent (.jar) to encode the main .bas, built with `is_encrypted` set to `true` and the encrypted bas. This works great.

The problem now is about the INCLUDE files. I had a look in `AddProgramLine.doInclude()` and if I understand correctly, it uses `Basic.getBufferedReader()`? So I actually would have to modify this method in order to do the same decryption as in `Basic.LoadTheProgram()`? Can you confirm?

**Comment by mougino** [comment]
I did that, modified `Basic.getBufferedReader()` the same way (so I took the decrypt method out in a `public static InputStream DecryptedStream(InputStream inputStream)` to capitalize on it), and the result works nicely. I was able to obfuscate the main .bas and an include .bas and have them both decrypted and run correctly in modified BASIC!

Now the hard part... get my Git environment clean and ready to be able to push the commits ;(

**Comment by mougino** [comment]
Marc, I committed the changes to my local repo (referenced above) but my local Git environment refuses to propose the pull to RFO-BASIC/Basic, I get:
```
Pull to https://github.com/RFO-BASIC/Basic.git
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'https://github.com/RFO-BASIC/Basic.git'
hint: Updates were rejected because a pushed branch tip is behind its remote
hint: counterpart. Check out this branch and integrate the remote changes
hint: (e.g. 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
No idea what it means..... If you can pull the changes manually from my repo that would be great, thanks!

**Comment by mougino** [comment]
Note that the "setup.xml" new boolean `is_encrypted` only impacts user APKs (`is_apk` needs to be set to true). For your peace of mind, you can integrate the new option in "setup.xml" with default value at false. Quick APK changes it to true if the user checks the option to encrypt his .bas files...

**Comment by jMarcS** [comment]
Hi, Nicolas,

Thanks for this new feature!

Your emails started arriving early in my work day. We had an all-hands
meeting, followed by an off-site luncheon for our retiring Vice-President
of Engineering. We gave him a 3D-printer as a going-away gift. This is
cool: the company that makes these printers uses their own printers to
manufacture parts for the printers!

So I have had no time to respond to you on GitHub. I'll go over there right
now and read what you've said.

- Marc

**Comment by jMarcS** [comment]
I think what happened is that your two commits went straight to the original repository. They are not committed in your fork. They are already on the HEAD of the RFO-BASIC/basic repo. Since your fork is behind the head of the original repo, a pull request makes no sense.

**Comment by mougino** [comment]
Oops :( sorry I did that, it never happened before. Feel free to rewrite some of my changes if they're not at your taste. I'll try to find how to push only on my repo for next time.

**Comment by jMarcS** [comment]
Your second commit is not complete, but I see what you're doing -- AddProgramLine.AddLine(String) will simply omit &nbsb with all the other spaces, non? They stay in the Editor text, but the interpreter never sees them.

I like that better than my idea. I was going to add a menu item that would change the text in the Editor. Let's do it your way.

(I am more than a day behind the forum again, so I do not know what you said there.

**Comment by jMarcS** [comment]
`Basic.getBufferedReader()` is used in three places.

Your change is for `AddProgramLine.doInclude()`.

The second use is in `Basic.loadProgramFileToList()`. That's loading a program file, just as `doInclude()` does.

The third use is in `TEXT.OPEN`, when opening a file for read.

If you have encrypted the program files in `assets`, will you also encrypt the data files? If you do, this change is okay. If you do not, `TEXT.OPEN` will try to decrypt a file that is not encrypted.

Have you tried a opening and reading a text file from `assets` in an encrypted APK?

**Comment by mougino** [comment]
Indeed my second commit for html white spaces was incomplete, thank you for reporting (another Git mystery). I pushed the rest of the commit this morning. As you guessed correctly, `AddProgramLine.AddLine(String)` internally converts &nbsb to a normal space (that is then discarded by the interpreter if outside a string, same as usual), the &nbsp stays in the Editor text, but the interpreter never sees it.

**Comment by mougino** [comment]
Third use of `Basic.getBufferedReader(String dir, String path)` is problematic. I have no definite way to know if a data/ resource is used with `TEXT.OPEN`, or *only* used with `TEXT.OPEN`. User might access the text file in binary mode too maybe. Or he might not access it at all and copy it, etc. So I better keep any other file than main bas and include bas (i.e. everything outside SOURCE_DIR) **not** encrypted.
Another commit coming your way.

**Comment by jMarcS** [comment]
Nicolas, you have inspired me! You know I do not like duplicate code, and it has always bothered me that so much of `executeENCRYPT()` and `executeDECRYPT` are identical. Now that you are using similar code in Basic.java, it makes sense to have a custom class that does that common work.

I called it `Encryption`. The constructor takes a mode (`Cipher.ENCRYPT_MODE` or `Cipher.DECRYPT_MODE`) and a password String. It throws an Exception if any of the setup fails or the `Cipher.init()` fails. It has a getter method to fetch the Cipher object.

Your `DecryptedStream()` method (which I renamed to fit the other stream methods) looks like this:

    public static InputStream getDecryptedStream(InputStream inputStream) {
        // Decrypt program that was encrypted with PBEWithMD5AndDES
        String PW = BasicContext.getPackageName();

        ByteArrayOutputStream byteBuffer = new ByteArrayOutputStream();
        final int bufferSize = 1024;
        byte[] dec = new byte[bufferSize];

        String Dest = "@@@@";
        try {
            int len = 0;
            while ((len = inputStream.read(dec)) != -1) {	// Get bytes from .bas file
                byteBuffer.write(dec, 0, len);
            }
            dec = byteBuffer.toByteArray();

            Cipher dcipher = new Basic.Encryption(Cipher.DECRYPT_MODE, PW).cipher();
            byte[] utf8 = dcipher.doFinal(dec);				// Decrypt
            Dest = new String(utf8, "UTF8");				// Encode bytes to UTF 8 to get a string
        } catch (Exception e) {
            Log.e(LOGTAG, "LoadTheProgram - error preparing decryption: " + e);
            return null;
        } finally {
            if (inputStream != null) {
                try { inputStream.close(); }
                catch (IOException e) {}
            }
        }
        if (Dest.equals("@@@@")) {
            Log.e(LOGTAG, "LoadTheProgram - bad decryption password");
            return null;
        }
        return new ByteArrayInputStream(Dest.getBytes());	// Change back decrypted string to input stream
    }

You can see that I added a `finally` clause. This method replaces its `inputStream` argument. The old `inputStream` should be closed before it is discarded.

**Comment by jMarcS** [comment]
I have a question about the code in my last post. You and I both carried over Paul's `if (Dest.equals("@@@@"))`, a test to see if the decryption did nothing. When I test the method, a bad password always gets an Exception from one of the `Cipher` methods, probably `doFinal()`. I don't know any way to run decryption on a string and have it come out unchanged.

I have not seen this extra check added in examples on the Web. I don't know if Paul put it in because he knew it was needed, or because he did not know if it was not needed. Sorry, that's pretty convoluted English.

But I would like to delete that extra code. In your testing, have you ever seen your "bad decryption password" error message?

**Comment by jMarcS** [comment]
There is a better way: `CipherInputStream`:

    public static InputStream getDecryptedStream(InputStream inputStream) throws Exception {
        // Decrypt program that was encrypted with PBEWithMD5AndDES
        String PW = BasicContext.getPackageName();
        Cipher cipher = new Basic.Encryption(Cipher.DECRYPT_MODE, PW).cipher();
        return new CipherInputStream(inputStream, cipher);
    }

This solves the stream closure error very neatly. But I don't have a good way to test it. Would you like me to commit this untested or send you the source file?

**Comment by mougino** [comment]
I will test it, but the main difference between bas obfuscation and BASIC! commands Encrypt/Decrypt is that I do not use the additional Base64 encryption in bas obfuscation (to avoid double the size of the file), did you take it into account?

**Comment by mougino** [comment]
You can commit your changes Marc, I trust you did the correct job. As for the "bad decryption password" error message I never encountered it, I share your analysis: I kept it for legacy reasons but I think we're safe to remove it.

**Comment by mougino** [comment]
Marc, if you want to do some tests on obfuscated bas with `is_apk` set to true, you can download Quick APK (PC) Java util to encrypt in PBEWithMD5AndDES: available at http://mougino.free.fr/RFO/quick-apk/pbe_md5_des.jar and invoke it with `java -jar pbe_md5_des.jar encrypt password path/to/myfile.ext`

**Comment by jMarcS** [comment]
Sorry to leave you hanging so long, Nicolas. I have had zero time since my last post. I should have just committed my files so you could work. I would rather commit experiments to my fork, but  I ran out of time. Bad planning!

About your Base64 question: no problem. Your DecryptedStream method creates a Cipher object, initializes it with a password, copies the stream to a byte buffer, and decrypts it with the Cipher object. In mine, I get the same Cipher object, then give it to the CipherInputStream wrapper. You read the CipherInputStream just the same as the stream it wraps. It delivers bytes as if it were decrypting on the fly. It's probably doing the same stuff your code did, but it hides all the details. I don't know if it buffers the entire file.

The BASIC! Encrypt command creates the same Cipher object, too, using a different password, then encodes the user's string to Base64, then ecrypts the Base64 string with the Cipher object (other way around for Decrypt). There are no streams involved.

I was looking at how to write this in Appendix D. I think the name "is_encrypted" is, ahem, too cryptic. I'm using "apk_programs_encrypted". Change it again if you have a better idea.

**Comment by jMarcS** [comment]
I forgot to put the Issue number in the commit comment. But the changes are available now, both in my fork and the RFO-BASIC repo.

**Comment by jMarcS** [comment]
Nicolas, I am not comfortable with this change in `Basic.java:getBufferedReader(String, String)`

    if (res.getBoolean(R.bool.apk_programs_encrypted) && dir == SOURCE_DIR)

It is not wrong, but perhaps too uncertain? I think that `getBufferedReader` called from `Run.java.executeTEXT_OPEN()` should never do decryption at all. I propose this change:

In `Basic.java:getBufferedReader`, add a Boolean parameter:

    -  public static BufferedReader getBufferedReader(String dir, String path) throws Exception {
    +  public static BufferedReader getBufferedReader(String dir, String path, boolean enableDecryption) throws Exception {

then inside the method, only if isAPK:

    -       if (res.getBoolean(R.bool.apk_programs_encrypted) && dir == SOURCE_DIR) {
    +       if (enableDecryption & res.getBoolean(R.bool.apk_programs_encrypted)) {

Where it is called:

    AddProgramLine.java:doInclude(String)
    -       try { buf = Basic.getBufferedReader(Basic.SOURCE_DIR, fileName); }
    +       try { buf = Basic.getBufferedReader(Basic.SOURCE_DIR, fileName, Basic.Encryption.ENABLE_DECRYPTION); }

    Basic.java:loadProgramFileToList(...)
    -       try { buf = getBufferedReader((isFullPath ? null : Basic.SOURCE_DIR), path); }
    +       String dir = isFullPath ? null : Basic.SOURCE_DIR;
    +       try { buf = getBufferedReader(dir, path, Basic.Encryption.ENABLE_DECRYPTION); }

    Run.java:executeTEXT_OPEN()
    -       buf = Basic.getBufferedReader(Basic.DATA_DIR, fileName);
    +       buf = Basic.getBufferedReader(Basic.DATA_DIR, fileName, Basic.Encryption.NO_DECRYPTION);

I believe this does what you need. `doInclude` and `loadProgramFileToList` always check `apk_programs_encrypted` -- I'm not sure that `loadProgramFileToList` always needs to, but the check does no harm. But `executeTEXT_OPEN()` never checks and never decrypts.

I am reluctant to push this on you, because it is only my "feeling". This is your feature, not mine. But if you do not object, I would like to make this change.

**Comment by mougino** [comment]
More than I don't object, I'm in favor ;)
You are the architect, Marc, this change doesn't break my feature and makes you more comfortable, I don't see why I could be against it...

**Comment by jMarcS** [comment]
Thank you, mon ami.

**Comment by jMarcS** [comment]
Released in v01.86, 2015/04/11.
